POC/ColorPicker/ColorPicker/ColorPickingFunctionality/SystemEvents/SystemHook.cs

```diff
@@ -0,0 +1,111 @@
+        /// <returns> 1 on success, 0 on failure. </returns>
+        [DllImport("user32.dll", SetLastError = true)]
+        private static extern int UnhookWindowsHookEx(int hookHandle);
+
+        /// <summary> 
+        /// This method calls the next hook in the chain. This allows the event to pass through if you do not want to sotp it. 
+        /// </summary>
+        /// <param name="hookHandle"> This is the handle of the hook to be removed. </param>
+        /// <param name="nCode"> The hook code passed to the current hook procedure. </param>
+        /// <param name="wParam"> Specifies whether the message is sent by the current process (Non-zero if from current process, NULL otherwise). </param>
+        /// <param name="lParam"> A pointer to a structure that contains details about the message. </param>
+        /// <returns> The value returned by the next hook in the chain. The current hook also needs to return this vlaue. </returns>
+        [DllImport("user32.dll")]
+        private static extern int CallNextHookEx(int hookHandle, int nCode, int wParam, IntPtr lParam);
+
+        /// <summary>
+        /// Retrieves a module handle for the specified module.
+        /// </summary>
+        /// <param name="moduleName"> The name of the loaded module (either a .dll or .exe file). </param>
+        /// <returns> A handle to specified file, or NULL on failure. </returns>
+        [DllImport("kernel32.dll")]
+        private static extern IntPtr GetModuleHandle(string moduleName);
+
+        public SystemHook(int hookType)
+        {
+            hookActionDelegate += HookProc;
+            this.hookType = hookType;
+            CaptureGlobalEvent();
+        }
+
+        ~SystemHook()
+        {
+            ReleaseGlobalEvent();
+        }
+
+        public abstract int HookProc(int nCode, int wParam, IntPtr lParam);
+
+        public int CallNextHookExWrapper(int nCode, int wParam, IntPtr lParam)
+        {
+            return CallNextHookEx(hookHandleID, nCode, wParam, lParam);
+        }
+
+        private void CaptureGlobalEvent()
+        {
+            if (hookHandleID == 0)
+            {
+                hookHandleID = SetWindowsHookEx(
+                    hookType,
+                    hookActionDelegate,
+                    SafeGetModuleHandle(Process.GetCurrentProcess().MainModule.ModuleName),
+                    0);
+                if (hookHandleID == 0)
+                {
+                    throw new Win32Exception(Marshal.GetLastWin32Error());
+                }
+            }
+        }
+
+        private IntPtr SafeGetModuleHandle(string moduleName)
+        {
+            IntPtr handle = GetModuleHandle(moduleName);
+            if (handle == IntPtr.Zero)
+            {
+                throw new InternalSystemCallException("Failed to get module handle");
+            }
+            return handle;
+        }
+
+        private void ReleaseGlobalEvent()
+        {
+            if (hookHandleID != 0)
```

Comment:
if (hookHandleID && !UnhookWindowsHookEx(hookHandleID))
{
    // code
}

May be preference but if the first condition fails in the if statement [hookHandle = 0], will exit and not check next condition - will "short circuit"